app/src/main/kotlin/compiler/analysis/GeneratorResolver.kt

```diff
@@ -0,0 +1,66 @@
+package compiler.analysis
+
+import compiler.ast.AstNode
+import compiler.ast.Expression
+import compiler.ast.Function
+import compiler.ast.Program
+import compiler.ast.Statement
+import compiler.ast.Variable
+import compiler.utils.Ref
+import compiler.utils.mutableKeyRefMapOf
+
+object GeneratorResolver {
+    fun computeGeneratorProperties(ast: Program): Map<Ref<Function>, List<Ref<Statement.ForeachLoop>>> {
+
+        val resultMapping: MutableMap<Ref<Function>, MutableList<Ref<Statement.ForeachLoop>>> = mutableKeyRefMapOf()
+
+        fun process(surroundingGenerators: List<Function>, node: AstNode) {
+            fun process(surroundingGenerators: List<Function>, vararg bunchOfBlocks: List<Statement>?) =
+                bunchOfBlocks.forEach { block -> block?.forEach { process(surroundingGenerators, it) } }
+
+            fun processFunction(surroundingGenerators: List<Function>, function: Function) {
+                val updatedSurroundingGenerators = if (function.isGenerator) surroundingGenerators + listOf(function) else surroundingGenerators
+                if (function.isGenerator) resultMapping.putIfAbsent(Ref(function), mutableListOf())
+                process(updatedSurroundingGenerators, function.body)
+            }
+
+            fun processForEachLoop(surroundingGenerators: List<Function>, foreachLoop: Statement.ForeachLoop) {
```

Comment:
This should be inlined as well, we only call this once.

Reply:
Inlined